### ISSUE TITLE
dependent: :delete_all only deletes direct children

### ISSUE THREAD

**Issue by yijia-zhan** [mceachen/closure_tree#242]
```ruby
class Tag
  has_closure_tree dependent: :delete_all
end
```
With the code above, `@tag.destroy` only deletes `@tag`'s direct children. Its grandchildren are all orphaned, keeping old `parent_id`s which no longer exist. However, the hierarchy records of all direct children and grandchildren do get deleted though. Is this intended? 

Thanks.

**Comment by mceachen** [comment]
Foreign keys shouldn't ever point to deleted records (and fk constraints
should prevent that from happening). It should be a recursive delete.

Thanks for reporting this. PRs are welcome!

**Comment by yijia-zhan** [comment]
Yeah, I'm surprised this can actually happen. Searched in the codebase a little bit, the `:dependent` option is simply only being [passed into ActiveRecord](https://github.com/mceachen/closure_tree/blob/2449029ecb7f287198154057bae88a912bfb5352/lib/closure_tree/model.rb#L21), `closure_tree` doesn't seem to be doing any particular about destroying.

**Comment by mceachen** [comment]
Yeah, my use case never deletes records (!), so it's not well exercised
except by tests. It should only be one extra query to join the hierarchy
table and get recursive delete to happen.

**Comment by mceachen** [comment]
If someone needs this and wants to take a swing at it, you'll need at least 3 steps:
1. capture all descendant IDs (either in a temp table or in memory) via the `hierarchy_` table
2. `DELETE` all the references from the `hierarchy_` table
3. `DELETE` from the model table `WHERE id IN (all descendant IDs)`
